src/main/kotlin/io/hsar/wh40k/combatsimulator/logic/TacticalActionStrategy.kt

```diff
@@ -1,67 +1,124 @@
+                                    listOf(targetedAction, anotherAction)
                                 }
-                        else -> eachTargetedAction to 0
-                    }
+                            }
+                            .flatten()
+                }
+                // Remove all half action combos where both actions are the same
+                .filterNot { (firstAction, secondAction) ->
+                    firstAction == secondAction
+                }
+                // Remove illegal combinations
+                .filter { halfActionCombo ->
+                    isLegalActionPair(halfActionCombo)
                 }
-                .maxBy { (_, expectedDamage) -> expectedDamage }
-                ?.first
 
-        return listOfNotNull(aimAction, maxDamageAttackAction)
+        return (fullActionTargetedActions + allLegalHalfActionCombos)
+                .map { eachLegalActionCombo ->
+                    getExpectedValue(world, thisUnit, eachLegalActionCombo) to eachLegalActionCombo
+                }
+                .maxBy { (expectedValue, _) -> expectedValue }!!
+                .second
     }
 
-    fun getPossibleTargetedActions(world: World, thisUnit: UnitInstance, possibleAttacks: List<DamageCausingAction>): List<TargetedAction> {
+    fun getExpectedValue(world: World, thisUnit: UnitInstance, actionCombo: List<TargetedAction>): Float {
+        return world.createCopy()
+                .let { tempWorld ->
+                    tempWorld.replaceUnitInstanceWithCopy(thisUnit)
+                            .let { tempUser ->
+                                actionCombo.map { targetedAction ->
+                                    // only replace target if not the same as user (eg for aim action) or the user reference will not line up
+                                    // and make sure to copy user for each action combo
+                                    if(thisUnit != targetedAction.target) {
+                                        tempWorld.replaceUnitInstanceWithCopy(targetedAction.target)
+                                    } else {
+                                        tempUser  // already temp as we have copied the user
+                                    }
+                                            // TODO not a massive issue but if both actions of a combo have the same target then any changes to the target won't carry over
+                                            .let { tempTarget ->
+                                                if (targetedAction.action.isLegal(tempWorld, tempUser, tempTarget)) {
+                                                    targetedAction.action.expectedValue(tempWorld, tempUser, tempTarget)
+                                                            .also {
+                                                                targetedAction.action.apply(tempWorld, tempUser, tempTarget)
```

Comment:
Just thinking, if `apply()` returns the new World state maybe we could get simplify (even if we can't get totally around) the copy problem.

Reply:
That sounds like a good idea in principle, although you'd have to deep copy world to return it or you'd have the same issue due to the obj references in world pointing to the same things.

Reply:
Perhaps could refactor in a future feature branch?